chandra_aca/centroid_resid.py

```diff
@@ -0,0 +1,325 @@
+
+        Along the way set self.agasc_id.  One can also just set
+        self.ra and dec directly.
+        """
+        if obsid is not None and slot is not None:
+            if date is None:
+                raise ValueError("Need date to look up star catalog if obsid/slot provided")
+            sc = mica.starcheck.get_starcheck_catalog_at_date(date)
+            star = sc['cat'][(sc['cat']['slot'] == slot)
+                             & ((sc['cat']['type'] == 'GUI') | (sc['cat']['type'] == 'BOT'))]
+            if not len(star):
+                raise ValueError(
+                    "No GUI or BOT in slot {} for obsid {} in dwell".format(slot, obsid))
+            agasc_id = star['id'][0]
+        # Could also add logic to infer star from loose position and magnitude
+        star = agasc.get_star(agasc_id, date=date)
+        self.agasc_id = agasc_id
+        self.ra = star['RA_PMCORR']
+        self.dec = star['DEC_PMCORR']
+
+    @property
+    def yags(self):
+        return self._yags
+
+    @yags.setter
+    def yags(self, vals):
+        if isinstance(vals, fetch.MSID):
+            self._yags = np.array(vals.vals)
+            self._yag_times = vals.times
+        else:
+            self._yags = np.array(vals)
+
+    @property
+    def yag_times(self):
+        return self._yag_times
+
+    @yag_times.setter
+    def yag_times(self, vals):
+        self._yag_times = np.array(vals)
+
+    @property
+    def zags(self):
+        return self._zags
+
+    @zags.setter
+    def zags(self, vals):
+        if isinstance(vals, fetch.MSID):
+            self._zags = np.array(vals.vals)
+            self._zag_times = vals.times
+        else:
+            self._zags = np.array(vals)
+
+    @property
+    def zag_times(self):
+        return self._zag_times
+
+    @zag_times.setter
+    def zag_times(self, vals):
+        self._zag_times = np.array(vals)
+
+    def set_offsets(self):
+        """
+        Apply time offsets to centroids based on type and source of centroid, obsid
+        (suggesting 8x8 or 6x6 data), and aspect solution source.  These time offsets were
+        fit.  The offsets determined against OBC aspect solutions are quite variable and
+        this code will warn when applying offsets in the case when residuals are to be
+        calculated using an OBC solution.
+        """
+
+        if self._dt_applied is True:
+            raise ValueError("Attempted to apply centroid dt more than once")
+        is_or = self.obsid < 38000
+        # Offsets calculated using fit scripts in SKA/analysis/centroid_and_sol_time_offsets
+        # Also see plots in PR https://github.com/sot/chandra_aca/pull/25
+        if self.att_source == 'obc':
+            warnings.warn("Applying centroid time offset for {} centroids and obc att source.".format(
+                self.centroid_source))
+        # For ground centroids we don't need the fetch source
+        if self.centroid_source == "ground":
+            if is_or:
```

Comment:
You should be able to make an inline table with 16 entries.
```
offsets = [# centroid_source, att_source, fetch_source, is_or, centroid_dt
           ['ground', 'ground', 'cxc', True, 0.125],
           ['ground', 'obc', 'cxc', True, 0.665],
...
]
offsets = {tuple(offset[:4]): offset[4] for offset in offsets}
```
It looks like there is an asymmetry here for caring about `fetch_source`, since it looks like a don't-care for `centroid_source = 'ground'`.  You only have 12 values here, but in fact you have plotted offsets for all 16 cases in the discussion.